sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1255 @@
+        self.coefs_ = []
+        self.intercepts_ = []
+
+        for i in range(self.n_layers_ - 1):
+            rng = check_random_state(self.random_state)
+            coef_init, intercept_init = self._init_coef(layer_units[i],
+                                                        layer_units[i + 1],
+                                                        rng)
+            self.coefs_.append(coef_init)
+            self.intercepts_.append(intercept_init)
+
+        if self.algorithm in _STOCHASTIC_ALGOS:
+            self.loss_curve_ = []
+            self._no_improvement_count = 0
+            if self.early_stopping:
+                self.validation_scores_ = []
+                self.best_validation_score_ = -np.inf
+            else:
+                self.best_loss_ = np.inf
+
+    def _init_coef(self, fan_in, fan_out, rng):
+        if self.activation == 'logistic':
+            # Use the initialization method recommended by
+            # Glorot et al.
+            init_bound = np.sqrt(2. / (fan_in + fan_out))
+        elif self.activation == 'tanh':
+            init_bound = np.sqrt(6. / (fan_in + fan_out))
+        elif self.activation == 'relu':
+            init_bound = np.sqrt(6. / (fan_in + fan_out))
+        else:
+            # this was caught earlier, just to make sure
+            raise ValueError("Unknown activation function %s" %
+                             self.activation)
+
+        coef_init = rng.uniform(-init_bound, init_bound, (fan_in, fan_out))
+        intercept_init = rng.uniform(-init_bound, init_bound, fan_out)
+        return coef_init, intercept_init
+
+    def _fit(self, X, y, incremental=False):
+        # Make sure self.hidden_layer_sizes is a list
+        hidden_layer_sizes = self.hidden_layer_sizes
+        if not hasattr(hidden_layer_sizes, "__iter__"):
+            hidden_layer_sizes = [hidden_layer_sizes]
+        hidden_layer_sizes = list(hidden_layer_sizes)
+
+        # Validate input parameters.
+        self._validate_hyperparameters()
+        if np.any(np.array(hidden_layer_sizes) <= 0):
+            raise ValueError("hidden_layer_sizes must be > 0, got %s." %
+                             hidden_layer_sizes)
+
+        X, y = self._validate_input(X, y, incremental)
+        n_samples, n_features = X.shape
+
+        # Ensure y is 2D
+        if y.ndim == 1:
+            y = y.reshape((-1, 1))
+
+        self.n_outputs_ = y.shape[1]
+
+        layer_units = ([n_features] + hidden_layer_sizes +
+                       [self.n_outputs_])
+
+        if not hasattr(self, 'coefs_') or (not self.warm_start and not
+                                           incremental):
+            # First time training the model
+            self._initialize(y, layer_units)
+
+        # l-bfgs does not support mini-batches
+        if self.algorithm == 'l-bfgs':
+            batch_size = n_samples
+        else:
+            if self.batch_size < 1 or self.batch_size > n_samples:
```

Comment:
optional: make operator precedence clear to those who don't know it by using parentheses